cl/api/utils.py

```diff
@@ -704,56 +705,58 @@ def invert_user_logs(
-                2016-10-01: 22,
-                total: 22,
-            },
-            joe_hazard: {
-                2016-10-01: 33,
-                total: 33,
-            }
-        }
-    :param start: The beginning date (inclusive) you want the results for. A
-    :param end: The end date (inclusive) you want the results for.
-    :param add_usernames: Stats are stored with the user ID. If this is True,
-    add an alias in the returned dictionary that contains the username as well.
-    :return The inverted dictionary
+    :return: Dictionary mapping user identifiers (usernames if `add_usernames=True`,
+        otherwise user IDs) to their daily API usage counts and totals.
+        Inner dictionaries are ordered by date. Only dates with usage are included.
     """
     r = get_redis_interface("STATS")
     pipe = r.pipeline()
 
     dates = make_date_str_list(start, end)
     for d in dates:
         pipe.zrange(f"api:v3.user.d:{d}.counts", 0, -1, withscores=True)
+        pipe.zrange(f"api:v4.user.d:{d}.counts", 0, -1, withscores=True)
+
+    # results contains alternating v3/v4 API usage data for each date queried.
+    # For example, if querying 2023-01-01 to 2023-01-02, results might look like:
+    # [
+    #     # 2023-01-01 v3 data: [(user_id, count), ...]
+    #     [("1", 100.0), ("2", 50.0)],
+    #     # 2023-01-01 v4 data
+    #     [("1", 50.0), ("2", 25.0)],
+    #     # 2023-01-02 v3 data
+    #     [("1", 200.0), ("2", 100.0)],
+    #     # 2023-01-02 v4 data
+    #     [("1", 100.0), ("2", 50.0)]
+    # ]
+    # We zip this with dates to combine v3/v4 counts per user per day
     results = pipe.execute()
 
-    # results is a list of results for each of the zrange queries above. Zip
-    # those results with the date that created it, and invert the whole thing.
     out: defaultdict = defaultdict(dict)
-    for d, result in zip(dates, results):
-        for user_id, count in result:
-            if user_id == "None" or user_id == "AnonymousUser":
-                user_id = "AnonymousUser"
-            else:
-                user_id = int(user_id)
-            count = int(count)
-            if out.get(user_id):
-                out[user_id][d] = count
-                out[user_id]["total"] += count
-            else:
-                out[user_id] = {d: count, "total": count}
+
+    def update_user_counts(_user_id, _count, _date):
+        user_is_anonymous = _user_id == "None" or _user_id == "AnonymousUser"
+        _user_id = "AnonymousUser" if user_is_anonymous else int(_user_id)
+        _count = int(_count)
+        out.setdefault(_user_id, OrderedDict())
+        out[_user_id].setdefault(_date, 0)
+        out[_user_id][_date] += _count
+        out[_user_id].setdefault("total", 0)
+        out[_user_id]["total"] += _count
+
+    for d, v3_data, v4_data in zip(dates, results[::2], results[1::2]):
+        for user_id, count in chain(v3_data, v4_data):
+            update_user_counts(user_id, count, d)
```

Comment:
I think we can use the `itertools.batched` function in the outer loop to process the data in chunks. This function iterates over the list, yielding tuples of items with a maximum size of `n`.

Here's the documentation for reference: [itertools.batched](https://docs.python.org/3/library/itertools.html#itertools.batched)

```suggestion
    for d, api_usage in zip(dates, batched(results, 2)):
        for user_id, count in chain(*api_usage):
            update_user_counts(user_id, count, d)
```

This approach enhances future maintainability. If a new API version (e.g., v5) is introduced, we simply need to adjust the batch size instead of modifying the slicing expressions for previous versions (v3, v4) and adding a new slicing expression for v5.

Reply:
@elisa-a-v looks great!